### ISSUE TITLE
Nelder Mead Parallelisation error

### ISSUE THREAD

**Issue by rpcarvs** [anyoptimization/pymoo#348]
The Nelder-Mead method returns an error when trying to run in parallel/different processes. I have tried multiprocess and multiprocessing. The problem happens after finishing the first gen/iter. I could somehow run in version 0.5.0 with multiprocess.

The error is: **cannot pickle 'generator' object**

And the relevant code lines:
```
class MyProblem(ElementwiseProblem):

    def __init__(self, **kwargs):
        super().__init__(n_var=16, n_obj=1, n_ieq_constr=0, xl=x0/8, xu=x0*8, **kwargs)

    def _evaluate(self, x, out, *args, **kwargs):
        out["F"] = external_func(x)

pool = multiprocess.Pool(5) 
runner = StarmapParallelization(pool.starmap)
problem = MyProblem(elementwise_runner=runner)
termination = get_termination("time", "00:35:00")
algorithm = NelderMead()
res = minimize(problem, algorithm, termination=termination, seed=1, verbose=True)
pool.close()
```

Pymoon 0.6.0
multiprocess 0.70.14
dill 0.3.6
Python 3.8.13

```

=================================================
n_gen  |  n_eval  |     f_avg     |     f_min    
=================================================
     1 |       10 |  0.0062304922 |  0.0061510627
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File <timed exec>:16

File .../site-packages/pymoo/optimize.py:67, in minimize(problem, algorithm, termination, copy_algorithm, copy_termination, **kwargs)
     64     algorithm.setup(problem, **kwargs)
     66 # actually execute the algorithm
---> 67 res = algorithm.run()
     69 # store the deep copied algorithm in the result object
     70 res.algorithm = algorithm

File .../site-packages/pymoo/core/algorithm.py:141, in Algorithm.run(self)
    139 def run(self):
    140     while self.has_next():
--> 141         self.next()
    142     return self.result()

File .../site-packages/pymoo/core/algorithm.py:161, in Algorithm.next(self)
    159 # call the advance with them after evaluation
    160 if infills is not None:
--> 161     self.evaluator.eval(self.problem, infills, algorithm=self)
    162     self.advance(infills=infills)
    164 # if the algorithm does not follow the infill-advance scheme just call advance
    165 else:

File .../site-packages/pymoo/core/evaluator.py:69, in Evaluator.eval(self, problem, pop, skip_already_evaluated, evaluate_values_of, count_evals, **kwargs)
     65 # evaluate the solutions (if there are any)
     66 if len(I) > 0:
     67 
     68     # do the actual evaluation - call the sub-function to set the corresponding values to the population
---> 69     self._eval(problem, pop[I], evaluate_values_of, **kwargs)
     71 # update the function evaluation counter
     72 if count_evals:

File .../site-packages/pymoo/core/evaluator.py:90, in Evaluator._eval(self, problem, pop, evaluate_values_of, **kwargs)
     87 X = pop.get("X")
     89 # call the problem to evaluate the solutions
---> 90 out = problem.evaluate(X, return_values_of=evaluate_values_of, return_as_dictionary=True, **kwargs)
     92 # for each of the attributes set it to the problem
     93 for key, val in out.items():

File .../site-packages/pymoo/core/problem.py:187, in Problem.evaluate(self, X, return_values_of, return_as_dictionary, *args, **kwargs)
    184     only_single_value = not (isinstance(X, list) or isinstance(X, np.ndarray))
    186 # this is where the actual evaluation takes place
--> 187 _out = self.do(X, return_values_of, *args, **kwargs)
    189 out = {}
    190 for k, v in _out.items():
    191 
    192     # copy it to a numpy array (it might be one of jax at this point)

File .../site-packages/pymoo/core/problem.py:227, in Problem.do(self, X, return_values_of, *args, **kwargs)
    225 # do the function evaluation
    226 if self.elementwise:
--> 227     self._evaluate_elementwise(X, out, *args, **kwargs)
    228 else:
    229     self._evaluate_vectorized(X, out, *args, **kwargs)

File .../site-packages/pymoo/core/problem.py:245, in Problem._evaluate_elementwise(self, X, out, *args, **kwargs)
    242 f = self.elementwise_func(self, args, kwargs)
    244 # execute the runner
--> 245 elems = self.elementwise_runner(f, X)
    247 # for each evaluation call
    248 for elem in elems:
    249 
    250     # for each key stored for this evaluation

File .../site-packages/pymoo/core/problem.py:37, in StarmapParallelization.__call__(self, f, X)
     36 def __call__(self, f, X):
---> 37     return list(self.starmap(f, [[x] for x in X]))

File .../site-packages/multiprocess/pool.py:372, in Pool.starmap(self, func, iterable, chunksize)
    366 def starmap(self, func, iterable, chunksize=None):
    367     '''
    368     Like `map()` method but the elements of the `iterable` are expected to
    369     be iterables as well and will be unpacked as arguments. Hence
    370     `func` and (a, b) becomes func(a, b).
    371     '''
--> 372     return self._map_async(func, iterable, starmapstar, chunksize).get()

File .../site-packages/multiprocess/pool.py:771, in ApplyResult.get(self, timeout)
    769     return self._value
    770 else:
--> 771     raise self._value

File .../site-packages/multiprocess/pool.py:537, in Pool._handle_tasks(taskqueue, put, outqueue, pool, cache)
    535     break
    536 try:
--> 537     put(task)
    538 except Exception as e:
    539     job, idx = task[:2]

File .../site-packages/multiprocess/connection.py:209, in _ConnectionBase.send(self, obj)
    207 self._check_closed()
    208 self._check_writable()
--> 209 self._send_bytes(_ForkingPickler.dumps(obj))

File .../site-packages/multiprocess/reduction.py:54, in ForkingPickler.dumps(cls, obj, protocol, *args, **kwds)
     51 @classmethod
     52 def dumps(cls, obj, protocol=None, *args, **kwds):
     53     buf = io.BytesIO()
---> 54     cls(buf, protocol, *args, **kwds).dump(obj)
     55     return buf.getbuffer()

File .../site-packages/dill/_dill.py:394, in Pickler.dump(self, obj)
    392 def dump(self, obj): #NOTE: if settings change, need to update attributes
    393     logger.trace_setup(self)
--> 394     StockPickler.dump(self, obj)

File .../pickle.py:487, in _Pickler.dump(self, obj)
    485 if self.proto >= 4:
    486     self.framer.start_framing()
--> 487 self.save(obj)
    488 self.write(STOP)
    489 self.framer.end_framing()

File .../site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File .../pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File .../pickle.py:901, in _Pickler.save_tuple(self, obj)
    899 write(MARK)
    900 for element in obj:
--> 901     save(element)
    903 if id(obj) in memo:
    904     # Subtle.  d was not in memo when we entered save_tuple(), so
    905     # the process of saving the tuple's elements must have saved
   (...)
    909     # could have been done in the "for element" loop instead, but
    910     # recursive tuples are a rare thing.
    911     get = self.get(memo[id(obj)][0])

File .../site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File .../pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File .../pickle.py:886, in _Pickler.save_tuple(self, obj)
    884 if n <= 3 and self.proto >= 2:
    885     for element in obj:
--> 886         save(element)
    887     # Subtle.  Same as in the big comment below.
    888     if id(obj) in memo:

File .../site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File .../pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File .../pickle.py:886, in _Pickler.save_tuple(self, obj)
    884 if n <= 3 and self.proto >= 2:
    885     for element in obj:
--> 886         save(element)
    887     # Subtle.  Same as in the big comment below.
    888     if id(obj) in memo:

File .../site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File .../pickle.py:603, in _Pickler.save(self, obj, save_persistent_id)
    599     raise PicklingError("Tuple returned by %s must have "
    600                         "two to six elements" % reduce)
    602 # Save the reduce() output and finally memoize the object
--> 603 self.save_reduce(obj=obj, *rv)

File .../pickle.py:717, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    715 if state is not None:
    716     if state_setter is None:
--> 717         save(state)
    718         write(BUILD)
    719     else:
    720         # If a state_setter is specified, call it instead of load_build
    721         # to update obj's with its previous state.
    722         # First, push state_setter and its tuple of expected arguments
    723         # (obj, state) onto the stack.

File .../site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File .../pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File .../site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File .../pickle.py:971, in _Pickler.save_dict(self, obj)
    968     self.write(MARK + DICT)
    970 self.memoize(obj)
--> 971 self._batch_setitems(obj.items())

File .../pickle.py:997, in _Pickler._batch_setitems(self, items)
    995     for k, v in tmp:
    996         save(k)
--> 997         save(v)
    998     write(SETITEMS)
    999 elif n:

File .../site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File .../pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File .../site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File .../pickle.py:971, in _Pickler.save_dict(self, obj)
    968     self.write(MARK + DICT)
    970 self.memoize(obj)
--> 971 self._batch_setitems(obj.items())

File .../pickle.py:1002, in _Pickler._batch_setitems(self, items)
   1000     k, v = tmp[0]
   1001     save(k)
-> 1002     save(v)
   1003     write(SETITEM)
   1004 # else tmp is empty, and we're done

File .../site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File .../pickle.py:603, in _Pickler.save(self, obj, save_persistent_id)
    599     raise PicklingError("Tuple returned by %s must have "
    600                         "two to six elements" % reduce)
    602 # Save the reduce() output and finally memoize the object
--> 603 self.save_reduce(obj=obj, *rv)

File .../pickle.py:717, in _Pickler.save_reduce(self, func, args, state, listitems, dictitems, state_setter, obj)
    715 if state is not None:
    716     if state_setter is None:
--> 717         save(state)
    718         write(BUILD)
    719     else:
    720         # If a state_setter is specified, call it instead of load_build
    721         # to update obj's with its previous state.
    722         # First, push state_setter and its tuple of expected arguments
    723         # (obj, state) onto the stack.

File .../site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File .../pickle.py:560, in _Pickler.save(self, obj, save_persistent_id)
    558 f = self.dispatch.get(t)
    559 if f is not None:
--> 560     f(self, obj)  # Call unbound method with explicit self
    561     return
    563 # Check private dispatch table if any, or else
    564 # copyreg.dispatch_table

File .../site-packages/dill/_dill.py:1186, in save_module_dict(pickler, obj)
   1183     if is_dill(pickler, child=False) and pickler._session:
   1184         # we only care about session the first pass thru
   1185         pickler._first_pass = False
-> 1186     StockPickler.save_dict(pickler, obj)
   1187     logger.trace(pickler, "# D2")
   1188 return

File .../pickle.py:971, in _Pickler.save_dict(self, obj)
    968     self.write(MARK + DICT)
    970 self.memoize(obj)
--> 971 self._batch_setitems(obj.items())

File .../pickle.py:997, in _Pickler._batch_setitems(self, items)
    995     for k, v in tmp:
    996         save(k)
--> 997         save(v)
    998     write(SETITEMS)
    999 elif n:

File .../site-packages/dill/_dill.py:388, in Pickler.save(self, obj, save_persistent_id)
    386     msg = "Can't pickle %s: attribute lookup builtins.generator failed" % GeneratorType
    387     raise PicklingError(msg)
--> 388 StockPickler.save(self, obj, save_persistent_id)

File .../pickle.py:578, in _Pickler.save(self, obj, save_persistent_id)
    576 reduce = getattr(obj, "__reduce_ex__", None)
    577 if reduce is not None:
--> 578     rv = reduce(self.proto)
    579 else:
    580     reduce = getattr(obj, "__reduce__", None)

TypeError: cannot pickle 'generator' object
```

**Comment by cyrilpic** [comment]
Hello,

I may have had a similar issue and I was able to find a work around. In the current code, the whole algorithm objects gets sent along to evaluate the problem on the different processes, which implies pickling it. Having the algorithm object in the objective function is probably rarely needed.

Here is my work around:
```python
class FixedElementwiseEvaluationFunction(ElementwiseEvaluationFunction):
    """FixedElementwiseEvaluationFunction

    Fixes pickling issues in multi-processing
    """
    def __init__(self, problem, args, kwargs) -> None:
        kwargs.pop('algorithm', None)
        super().__init__(problem, args, kwargs)

...

problem = MyProblem(elementwise_runner=runner, elementwise_func=FixedElementwiseEvaluationFunction)
```

I hope this helps.

**Comment by rpcarvs** [comment]
Hey @cyrilpic,

That did the trick! Thanks!

**Comment by cyrilpic** [comment]
Glad it works! I would leave the issue open though, as I think this should probably be fixed in pymoo itself.

**Comment by rpcarvs** [comment]
Good idea.

**Comment by blankjul** [comment]
Good catch! In fact, I was thinking about including a lot more `generators` into algorithms. But being limited using `pickle` is a true downside. 

For some problems, it turned out to be useful to have the `algorithm` object available. For instance, (I know this is a hacky solution, but anyway) to activate or deactivate variables in an algorithm or to even map variables to subspaces. 

Have you considered using `ThreadPool`? That works without pickle.

**Comment by cyrilpic** [comment]
In a distributed setting, changes made to `algorithm` are not propagated back to the original object. That could cause some confusion. And in my case, it was a real performance killer and it took me a while to understand why that was happening.

Why not having some sort of flag in `Problem` (`needs_algorithm` or whatever) that controls this behavior?

**Comment by blankjul** [comment]
@cyrilpic Can you check the commit above? what do you think? 

I am not sure if the flag should be by default `False` or `True`, though.

**Comment by cyrilpic** [comment]
Commit looks great, thanks! Straightforward and easy to set. My guess is that using algorithm in problem is less common. So, defaulting the flag to `False` seems reasonable.

**Comment by M4a1x** [comment]
@cyrilpic your workaround is awesome. Thanks a ton! Also thanks to @blankjul for the fix! 

I'll try to briefly describe my Problem here in case someone looks through the issues with the same Problem as me.

I spent almost three full days on multiprocessing issues. I was trying to run the `minimize()` from a `tkinter` GUI in a separate process (to keep the GUI responsive) that in turn uses the `starmap` parallelization technique with processes as described in the docs. 

Unfortunately, if you combine this with a `multiprocessing.Queue` in the callback (to plot the population in realtime in the GUI) it works only if the queue stays in the main `tkinter` process and the pymoo process that called `minimize()`. 

Passing around the algorithm object (as described above) also passed around the queue, breaking the program with varying error messages. 

Usually either 
- `can't pickle <xxx>` or 
- `RuntimeError: Queue objects should only be shared between processes through inheritance`
would show up. I stumbled upon this issue as I noticed that they would print as many errors as I have processes running, seemed odd, as the callback is only called once after every generation.

@blankjul Would it be possible to get a hotfix release that includes commit feb31d0? Maybe embedding pymoo in tkinter+matplotlib would be also worth a consideration for the documentation?
